subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/CleanupStaleOutputsStepTest.groovy

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps
+
+import org.gradle.api.internal.file.TestFiles
+import org.gradle.internal.execution.BuildOutputCleanupRegistry
+import org.gradle.internal.execution.OutputChangeListener
+import org.gradle.internal.execution.UnitOfWork
+import org.gradle.internal.execution.history.OutputFilesRepository
+import org.gradle.internal.file.Deleter
+import org.gradle.internal.file.TreeType
+import spock.lang.Shared
+
+class CleanupStaleOutputsStepTest extends StepSpec<WorkspaceContext> {
+    def cleanupRegistry = Mock(BuildOutputCleanupRegistry)
+    def deleter = Mock(Deleter)
+    def outputChangeListener = Mock(OutputChangeListener)
+    def outputFilesRepository = Mock(OutputFilesRepository)
+
+    @Shared
+    def existing = file("existing.txt").createFile()
+    @Shared
+    def missing = file("missing.txt")
+
+    def step = new CleanupStaleOutputsStep<>(
+        buildOperationExecutor,
+        cleanupRegistry,
+        deleter,
+        outputChangeListener,
+        outputFilesRepository,
+        delegate)
+
+    def delegateResult = Mock(Result)
+
+    @Override
+    protected WorkspaceContext createContext() {
+        Stub(WorkspaceContext)
+    }
+
+    def "#description is cleaned up: #cleandUp"() {
+        def target = file("target")
+        creator(target)
+
+        when:
+        def result = step.execute(work, context)
+
+        then:
+        result == delegateResult
+
+        _ * work.shouldCleanupStaleOutputs() >> true
+        _ * work.visitOutputs(_) { UnitOfWork.OutputVisitor visitor ->
+            visitor.visitOutputProperty("output", TreeType.FILE, target, TestFiles.fixed(target))
+        }
+
+        _ * cleanupRegistry.isOutputOwnedByBuild(target) >> ownedByBuild
+        _ * outputFilesRepository.isGeneratedByGradle(target) >> generatedByGradle
+
+        then:
+        if (cleandUp) {
```

Comment:
💄 Variable name typo

Reply:
Thanks, fixed. Wonder why IntelliJ doesn't do spellchecks in `where:` variable names...